### PR TITLE
Update wasmtime dependency from 18 to 21

### PR DESCRIPTION
**Pull request by FrankReh**
Update the wasmtime dependency version from 18 to 21.

Also add type checks for the wasmtime/component.

Removed the `ty` method from the `WasmValue` trait.

Also add type checks for the value::{Value, Type} as the `ty` method is no longer available for building up a tree of types from a tree of values.

### CODE REVIEW

File: src/wasmtime/component.rs
Comment (lann) [inline src/wasmtime/component.rs]: This is one of the things I was expecting to need a trait redesign to handle.

File: src/wasmtime/core.rs
Comment (lann) [inline src/wasmtime/core.rs]: I think this can be handled with a `match self`. There are `const`s for ref types (which I haven't really wrapped my head around yet), e.g. https://docs.rs/wasmtime/20.0.2/wasmtime/struct.RefType.html#associatedconstant.EXTERNREF

Comment (FrankReh) [reply src/wasmtime/core.rs]: Nice. At least the names match up and they are constants. See the next commit.

File: src/wasmtime/component.rs
Comment (FrankReh) [reply src/wasmtime/component.rs]: So the tests don't run into the unimplemented cases. The `write` call uses the new `kind` method directly.

Maybe the idea of going to a Type can be changed and the logic can go directly to a Kind. That would be a more dramatic change.

Do you want to hold off reviewing the other changes until this is cleaned up? I drew inspiration from the original having used unimplemented for the more complicated types too. And do you want to do this now or is there time for me to try and figure out some more things?

Something that could help me if you think there is benefit in me trying to resolve this one, are there other unit tests you would like to seen work that aren't there yet?

Comment (lann) [reply src/wasmtime/component.rs]: Sorry I haven't had a chance to look at exactly what is going on here; I'll try to find time tomorrow.

Comment (FrankReh) [reply src/wasmtime/component.rs]: When you do look, notice that the use of val.ty().kind() has been replaced with val.kind() in two places - I think that is the interesting part. So the ty() trait method wasn't in play in the two critical areas.

Comment (lann) [reply src/wasmtime/component.rs]: OK, I have a better grasp on the problem now; here's what I'm thinking:

First, replace `WasmValue::ty` with `::kind`. As you note, some uses of `ty` really only need the kind.

The remaining uses of `ty` are for type checking, which this PR works around by relaxing those checks to only validate `kind`. I'd like to get that type checking back by adding a new method to `WasmValue`: `fn is(&self, ty: &Self::Type) -> bool`. I believe it should be relatively straightforward to implement this for `component::Val` though it may take a fair bit of code to handle all the composite types.

Does that sound reasonable? If you are willing to attempt the above in this PR that would be great, otherwise I'm fine merging this as-is and will probably try to make the other changes before the next release.

Comment (FrankReh) [reply src/wasmtime/component.rs]: > The remaining uses of ty are for type checking, which this PR works around by relaxing those checks to only validate kind. 

While I don't know *exactly* what you mean that some checks are relaxed. If you mean some calls to `ty` for a component::Val would result in a panic with the current PR commits, I agree. I wasn't sure if they were important to handle because I didn't see a unit test fail.

Would it take new unit tests to show them fail or some other API for the wasm-wave library itself that isn't fleshed out yet?

I would rather take on this next step if I could see it actually passing a test as a result.

So I can start trying to figure out the next step from your guidance, but if the current PR doesn't make things worse, I would prefer this go in, and other improvements go into a new PR. Makes understanding and possible git bisections later easier.

File: src/wasmtime/component.rs
Comment (lann) [inline src/wasmtime/component.rs]: Here is an example of the type checking being relaxed: [`Flags::new_val`](https://docs.rs/wasmtime/19.0.2/wasmtime/component/types/struct.Flags.html#method.new_val) which was called at 195 would validate that `names` matched the defined flag names. The same applies to many of the `make_*` diffs above.

File: src/wasmtime/component.rs
Comment (lann) [reply src/wasmtime/component.rs]: > I don't know exactly what you mean that some checks are relaxed. 

https://github.com/lann/wasm-wave/pull/34/files#r1601813873

Comment (lann) [reply src/wasmtime/component.rs]: The test coverage you're looking for is missing because the test cases can't be constructed just from the wasmtime API; you have to load an actual wasm file and reflect on its types to get the relevant test fixtures.

Comment (lann) [reply src/wasmtime/component.rs]: ...though immediately after typing that I remembered that I implemented exactly that test fixture for `tests/wasmtime.rs`, so I probably can just expand coverage there.

File: src/wasmtime/component.rs
Comment (FrankReh) [reply src/wasmtime/component.rs]: Oh, I was totally on the wrong track about what you were referring to.

Thanks for pointing me at this. These kinds of changes/regressions I would want to correct before the PR lands.

I can work backwards from a stack trace to see when these are being called when the crate is tested and see how more tests would have caught the regression.

File: src/wasmtime/component.rs
Comment (lann) [reply src/wasmtime/component.rs]: Minimal test case expansion which passes on `main` and fails on this PR: https://github.com/lann/wasm-wave/pull/35

Comment (lann) [reply src/wasmtime/component.rs]: I'd still be fine merging this PR as-is if you prefer, we'd just need to coordinate on adding that test coverage PR if you want to work on the followup.

Comment (FrankReh) [reply src/wasmtime/component.rs]: Thanks for getting back to me on this. I've been following the idea that a bad value for a Flags type was being caught with wasmtime 18 and now it isn't, at least the way I had changed things.

And sure enough, a test based on the tests/wasmtime/test_round_trips_unchanged does fail with I give it a garbage flag name. I'm in the process of figuring out how the wasmtime library knew that was not valid.

Comment (FrankReh) [reply src/wasmtime/component.rs]: Okay. I found the wasmtime 20 library has a 'names()' method on the Flags type. So it would be straightforward to perform the check manually in that make_flags method.

Two questions. Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information. And I wonder that the wasmtime library doesn't provide the validation check any longer. Does its author want each embedder to code their own checks?

---

Maybe that's what the comment from 3 months ago was referring to. The wasmtime library doesn't catch errors when things are created, only when they are attempted to be passed to the component (if I'm remembering that correctly).

Comment (lann) [reply src/wasmtime/component.rs]: Most wasmtime embedders use [`bindgen!`](https://docs.rs/wasmtime/20.0.2/wasmtime/component/macro.bindgen.html) which enforces types through code generation; they don't use this Val interface at all. I believe the API change allowed the implementation to be simplified quite a bit at a (small) expense to weird crates like this one :slightly_smiling_face:

File: src/wasmtime/component.rs
Comment (FrankReh) [reply src/wasmtime/component.rs]: Interesting. I see the merge of the make_list test. That's confirming the composite types match up.

Wonder if there is a way to make the inputs and outputs all strings, like the other tests. Or will these require coded calls to make_* for any valid and invalid combinations. Guess they can still be put into a table of inputs and expected type results.

Comment (FrankReh) [reply src/wasmtime/component.rs]: How about added checks along these lines? (Note the recursion in the second function.)
```rust
    fn make_list(
        ty: &Self::Type,
        vals: impl IntoIterator<Item = Self>,
    ) -> Result<Self, WasmValueError> {
        ensure_type_kind(ty, WasmTypeKind::List)?;
        let component::Type::List(list) = ty else {
            // TBD this could be where the error is returned rather
            // than having to make the call to ensure_type_kind.
            panic!();
        };
        let elem_type = list.ty();
        vals.into_iter()
            .map(|val| {
                is_value_type(&val, &elem_type)?;
                Ok(val)
            })
            .collect::<Result<Vec<_>, WasmValueError>>()
            .map(Self::List)
    }
```
and
```rust
fn is_value_type(val: &component::Val, ty: &component::Type) -> Result<(), WasmValueError> {
    if val.kind() != ty.kind() {
        return Err(WasmValueError::wrong_value_type(ty, val));
    }
    match val {
        component::Val::List(vals) => {
            let component::Type::List(list) = ty else {
                unreachable!(); // kind to kind check should make this unreachable
            };
            let elem_type = list.ty();
            for val in vals {
                is_value_type(val, &elem_type)?;
            }
        },
        component::Val::Record(_vec_string_val) => todo!(),
        component::Val::Tuple(_vec_val_) => todo!(),
        component::Val::Variant(_string, _option_box_val__) => todo!(),
        component::Val::Enum(_string) => todo!(),
        component::Val::Option(_option_box_val__) => todo!(),
        component::Val::Result(_result_option_box_val_option_box_val) => todo!(),
        component::Val::Flags(_vec_string_) => todo!(),
        component::Val::Resource(_resourceany) => todo!(),

        // Any leaf variant type was already tested with the kind comparisons above.
        _ => {},
    }
    Ok(())
}
```
I don't like the almost duplication of code, but in the first case, I wanted to collect the values and in the second case I didn't. Open to ideas...

Comment (FrankReh) [reply src/wasmtime/component.rs]: I just had a thought, it could save a lot of code if the make_list creates the List but before returning it, checks that it ends up being a matching type. If the time spent and heap allocations made for the failure case is acceptable in this scenario.

Comment (FrankReh) [reply src/wasmtime/component.rs]: I haven't added any tests. I can understand why there weren't tests for this verify issue before because the verify was being done by the `wasmtime` library, not this one. Now that this is responsible for verifying variables are being created correctly, some more negative tests may be in order.

File: src/wasmtime/component.rs
Comment (lann) [reply src/wasmtime/component.rs]: Sorry, realized I didn't respond to one of your questions:

> Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information.

I'm fine with either in this PR; nicer errors can always be added later.

File: src/wasmtime/component.rs
Comment (lann) [inline src/wasmtime/component.rs]: Wasmtime conveniently provides [helpers](https://docs.rs/wasmtime/20.0.2/wasmtime/component/enum.Type.html#method.unwrap_list) for these assertions:

```suggestion
            let list_handle = ty.unwrap_list();
```

File: src/wasmtime/component.rs
Comment (lann) [inline src/wasmtime/component.rs]: I think it would be better to remove this method from the trait entirely rather than allow it to panic. Would that work with the changes in this PR?

**Comment by FrankReh** [comment]
Well, in addition to wondering if the direction of the last commit suits you. I was wondering if you wanted to revisit the `Value::make_*` functions in general. They are in src/value/mod.rs. I can't find that they are used except from tests, with the exception of three: Value::make_list, Value::make_option, and Value::make_result, used in src/value/convert.rs, in supporting `From` traits. So arrays, vectors, options, and results, can be converted into `Value` types.

But they all generate their `Type` information automatically, stored as a local variable `ty`, which is then passed to the `make_*` function. So the type isn't going to be compared to anything that has come from a WIT definition.

And what would trigger creation of `Value` records, variants, enums or flags.

Also wondering if the `ty` field is needed in structs in src/value/mod.rs, unless maybe a value of this type could be created with an empty vec and then operations were planned that would allow adding to the list later?
```rust
#[derive(Debug, Clone, PartialEq)]
#[doc(hidden)]
pub struct List {
    ty: Arc<ListType>,
    elements: Vec<Value>,
}
```

I know for sure I don't see all the ways this wasm-wave library was planned to be used in the future, so its current form is just a stepping stone.

Thanks.

**Comment by FrankReh** [comment]
@lann What should we do? Want me to drop this line?

**Comment by lann** [comment]
Ah sorry, I lost track of your comment over the weekend.

> wondering if the direction of the last commit suits you

Yes, that looks good to me.

> I was wondering if you wanted to revisit the Value::make_* functions in general.

These are used here, which is how the parser makes typed values: https://github.com/lann/wasm-wave/blob/e69b66e10383295debd16ce4bd031620fc5e1c22/src/ast.rs#L207-L236

**Comment by FrankReh** [comment]
Still regarding the `value` module.

I understand that for the unit tests, the types represented by the `ty` variable passed through the call chain would come from the unit test functions themselves. But that isn't helping me understand where they would come from when this is used as a library. If a WIT file or files were being used, wouldn't the library be able to use the wasmtime core or component types?

I don't yet see the need for duplicating the logic behind building variables and checking types for the concrete wasmtime core and component branches, with the library proprietary `value::{Value, Type}` variables and types.

You can always take this over if you think that is easier than getting me to see this. Thanks.

**Comment by lann** [comment]
> wouldn't the library be able to use the wasmtime core or component types

That may be possible with the wasmtime 20 val type (it wasn't before), but there are some use cases that don't otherwise need to use wasmtime and it would represent a very heavyweight dependency just to get these value types. If these types could be moved into a 3rd party crate to be a shared dependency I think this would be feasible, but the details of resource types make that difficult, currently.

> You can always take this over if you think that is easier than getting me to see this.

I'm happy to if you'd prefer. What is the status of this PR as-is? Is it complete enough to merge?

**Comment by FrankReh** [comment]
Your comment reminded me that the wit-parser and wasmtime dependencies are optional.

**Comment by FrankReh** [comment]
The last commit has the `Tuple` checking arm fleshed out. There are a lot of types being checked.

**Comment by FrankReh** [comment]
@lann I think now it's ready.

Down the road, perhaps the check_type and check_type2 that are called in value/mod.rs will be called at a higher spot so they are only called once for any Node that is constructed. As it stands right now, the pieces are built up from leaf nodes and their types are also checked and then as higher order nodes are built, their types are checked again, going all the way down to the leaves. It seems unnecessary but that can wait.

Also, it seems the Record type in the value module isn't as flexible as the one in the wasmtime component module. The field names are not stored so there is no chance to provide the fields in a different order and no chance to omit optional fields. But maybe that's intentional.

**Review by lann — APPROVED** [review]
This looks great! Thanks for sticking though all that iteration!

**Pull request merged**
